sentry/src/routes/analytics.rs

```diff
@@ -0,0 +1,236 @@
+
+        if !valid_event_types.iter().any(|e| *e == &self.event_type[..]) {
+            Err(ResponseError::BadRequest(format!(
+                "invalid event_type, possible values are: {}",
+                valid_event_types.join(" ,")
+            )))
+        } else if !valid_metric.iter().any(|e| *e == &self.metric[..]) {
+            Err(ResponseError::BadRequest(format!(
+                "invalid metric, possible values are: {}",
+                valid_metric.join(" ,")
+            )))
+        } else if !valid_timeframe.iter().any(|e| *e == &self.timeframe[..]) {
+            Err(ResponseError::BadRequest(format!(
+                "invalid timeframe, possible values are: {}",
+                valid_timeframe.join(" ,")
+            )))
+        } else {
+            Ok(())
+        }
+    }
+}
+
+fn default_limit() -> u32 {
+    100
+}
+
+fn default_event_type() -> String {
+    "IMPRESSION".into()
+}
+
+fn default_metric() -> String {
+    "eventCounts".into()
+}
+
+fn default_timeframe() -> String {
+    "hour".into()
+}
+
+pub async fn publisher_analytics<A: Adapter>(
+    req: Request<Body>,
+    app: &Application<A>,
+) -> Result<Response<Body>, ResponseError> {
+    process_analytics(req, app, false, false)
+        .await
+        .map(success_response)
+}
+
+pub async fn analytics<A: Adapter>(
+    req: Request<Body>,
+    app: &Application<A>,
+) -> Result<Response<Body>, ResponseError> {
+    let request_uri = req.uri().to_string();
+    let redis = app.redis.clone();
+
+    match redis::cmd("EXISTS")
+        .arg(&request_uri)
+        .query_async::<_, String>(&mut redis.clone())
+        .await
+    {
+        Ok(response) => Ok(success_response(response)),
+        _ => {
+            let cache_timeframe = match req.extensions().get::<RouteParams>() {
+                Some(_) => 600,
+                None => 300,
+            };
+            let response = process_analytics(req, app, false, true).await?;
+            cache(&redis.clone(), request_uri, &response, cache_timeframe).await;
+            Ok(success_response(response))
+        }
+    }
+}
+
+pub async fn advertiser_analytics<A: Adapter>(
+    req: Request<Body>,
+    app: &Application<A>,
+) -> Result<Response<Body>, ResponseError> {
+    process_analytics(req, app, true, true)
+        .await
+        .map(success_response)
+}
+
+pub async fn process_analytics<A: Adapter>(
+    req: Request<Body>,
+    app: &Application<A>,
+    is_advertiser: bool,
+    skip_publisher_filter: bool,
+) -> Result<String, ResponseError> {
+    let query = serde_urlencoded::from_str::<AnalyticsQuery>(&req.uri().query().unwrap_or(""))?;
+    query.is_valid()?;
+
+    let applied_limit = cmp::min(query.limit, 200);
+    let (interval, period) = get_time_frame(&query.timeframe);
+    let time_limit = Utc::now().timestamp() - period;
+    let sess = req.extensions().get::<Session>();
+
+    let mut where_clauses = vec![format!("created > to_timestamp({})", time_limit)];
```

Comment:
I would like to keep all querying stuff separated in the db module.

Reply:
Sounds okay